util/random/rand.go

```diff
@@ -71,8 +76,60 @@ func (r *randstream) XORKeyStream(dst, src []byte) {
 	}
 }
 
-// New returns a new cipher.Stream that gets random data from Go's crypto/rand package.
+// READER_BYTES is how many bytes we expect from each source
+const READER_BYTES = 32
+
+// MixedrandHash is the hash used in XORKeyStream to mix all entropy sources together.
+// sha512 is chosen as its digest is 64 bytes long, matching the length of blake2xb's seed.
+var MixedrandHash = sha512.New
+
+type mixedrandstream struct {
+	Readers []io.Reader
+}
+
+func (r *mixedrandstream) XORKeyStream(dst, src []byte) {
+
+	l := len(dst)
+	if len(src) != l {
+		panic("XORKeyStream: mismatched buffer lengths")
+	}
+
+	// try to read 32 bytes from all readers and write them in a buffer
+	var b bytes.Buffer
+	var nerr int
+	buff := make([]byte, READER_BYTES)
+	for _, reader := range r.Readers {
+		n, err := io.ReadFull(reader, buff)
+		if err != nil {
+			nerr++
+		}
+		b.Write(buff[:n])
+	}
+
+	// we are ok with few sources being insecure (i.e., providing less than
+	// READER_BYTES bytes), but not all of them
+	if nerr == len(r.Readers) {
+		panic("all readers failed")
+	}
+
+	// create the XOF output, with hash of collected data as seed
+	h := MixedrandHash()
+	h.Write(b.Bytes())
+	seed := h.Sum(nil)
+	blake2 := blake2xb.New(seed)
+	blake2.XORKeyStream(dst, src)
+}
+
+// New returns a new cipher.Stream that gets random data from the given
+// readers. If no reader was provided, Go's crypto/rand package is used.
+// In order to use every source, it tries to read 32 bytes from each, and mix it
```

Comment:
Change to: Otherwise, for each source, 32 bytes are read. They are concatenated and then hashed, and the resulting hash is used as a seed to a PRNG.

(I don't think the callers need to know about all the details, they can read the source to know.)